Feature/multi sign

houdini91 opened the pull request:
Hi.
I have created a PR that includes a small refactor to the interfaces and the internal code.
1) Added support for multiple signature validation (Signature support was already implemented)
2) Signer/Verifier interfaces change suggestion to support both public,keyid methods.
3) Signer/Verifier interface change suggestion - remove keyid from the `sign`, `verify` methods..
4) Add/Fix tests accordingly 

After implementing the multi signature validation struct it seemed that maybe i can clean it up a bit by unified the multi signature flow and the current single signature validation flow.

shibumi commented on dsse/sign.go:
```diff
@@ -77,16 +78,27 @@ type SignVerifier interface {
 
 // EnvelopeSigner creates signed Envelopes.
-type EnvelopeSigner struct {
+type envelopeSigner struct {
```
Is there a reason why we unexport this type suddenly?

adityasaky replied:
I _think_ it's okay given the `New...` API, but I don't have a strong opinion here. cc @houdini91

adityasaky replied:
Hmm, I think I'm now leaning towards keeping this exported @houdini91.

adityasaky reviewed (commented):
I'd like to hear from others as well before we proceed with this one.

cc @dlorenc @marklodato @SantiagoTorres @mnm678

adityasaky commented on dsse/verify.go:
```diff
@@ -1,70 +1,116 @@
 
-			verified = true
+			// See https://github.com/in-toto/in-toto/pull/251
+			if val, ok := usedKeyids[keyID]; ok {
```
Also mitigates (accidental or otherwise) padding of acceptedKeys if I'm not mistaken

mnm678 commented on dsse/verify.go:
```diff
@@ -1,70 +1,116 @@
+			if s.KeyID != "" && keyID != "" && err == nil && s.KeyID != keyID {
+				continue
+			}
 			if err != nil {
```
should this check come before the if statement above it?

mnm678 commented on dsse/verify.go:
```diff
@@ -1,70 +1,116 @@
 }
 
-func (ev *EnvelopeVerifier) Verify(e *Envelope) error {
+type AcceptedKeys struct {
```
I think this would more accurately be 'AcceptedKey'

mnm678 commented on dsse/verify.go:
```diff
@@ -1,70 +1,116 @@
-			verified = true
+			// See https://github.com/in-toto/in-toto/pull/251
+			if val, ok := usedKeyids[keyID]; ok {
+				fmt.Printf("Found envelope signed by different subkeys of the same main key, Only one of them is counted towards the step threshold, KeyID=%s\n", val)
```
I think this needs a continue statement as well so that the key is not added to `acceptedKeys`? Or at least a test to make sure it skips the duplicate.

mnm678 reviewed (commented):
This looks like a good change overall, I left a few comments

houdini91 commented:
@mnm678 @adityasaky 
> I think this needs a continue statement as well so that the key is not added to acceptedKeys? Or at least a test to make sure it skips the duplicate.

Added both continue and tests

> should this check come before the if statement above it?

Moved error check above, 
Note i added a default keyid calculation using the public key (SHA256) when a keyid is empty.
If a keyid is not found or calculated the empty keyid is used (Only one Verifier may use it)

adityasaky reviewed (approved):
I think this is fine. @mnm678 could I trouble you for another quick look? :smile:

mnm678 reviewed (approved):
LGTM with the one typo fix

houdini91 commented:
@adityasaky @mnm678 
I added a small change to prevent logs posted on already verified providers.
The problem was that all providers where checked against all signatures even when a provider already has verified a previous signature.
I am removing any verified provider from the next signature verification process.
Meaning a provider can only verify one signature.
Do you have any opinion on this does this seem logical?

adityasaky commented on dsse/verify.go:
```diff
@@ -1,70 +1,142 @@
+	return fingerprint, nil
+}
+
+func RemoveIndex(v []Verifier, index int) []Verifier {
```
```suggestion
func removeIndex(v []Verifier, index int) []Verifier {
```

adityasaky reviewed (commented):
>  I am removing any verified provider from the next signature verification process.

This sounds reasonable to me. @mnm678?

Also, let's make envelopeSigner public.

mnm678 commented:
> This sounds reasonable to me. @mnm678?

Makes sense to me. The duplicates are also removed before checking the threshold, but this does make the logs clearer.

houdini91 commented:
> Also, let's make envelopeSigner public (it's used in in-toto-golang, for example).
Sorry for the delay, changed.
@adityasaky

adityasaky commented:
Thanks, @houdini91!
